mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala

```diff
@@ -0,0 +1,238 @@
+   * Using a piecewise constant function
+   *
+   * @param testData features to be labeled
+   * @return predicted labels
+   */
+  def predict(testData: JavaRDD[java.lang.Double]): JavaDoubleRDD =
+    JavaDoubleRDD.fromRDD(predict(testData.rdd.asInstanceOf[RDD[Double]]))
+
+  /**
+   * Predict a single label
+   * Using a piecewise constant function
+   *
+   * @param testData feature to be labeled
+   * @return predicted label
+   */
+  def predict(testData: Double): Double = {
+    val result = binarySearch(features, testData)
+
+    val index =
+      if (result == -1) {
+        0
+      } else if (result < 0) {
+        -result - 2
+      } else {
+        result
+      }
+
+    labels(index)
+  }
+}
+
+/**
+ * Isotonic regression
+ * Currently implemented using parallel pool adjacent violators algorithm
+ */
+class IsotonicRegression
+  extends Serializable {
+
+  /**
+   * Run algorithm to obtain isotonic regression model
+   *
+   * @param input (label, feature, weight)
+   * @param isotonic isotonic (increasing) or antitonic (decreasing) sequence
+   * @return isotonic regression model
+   */
+  def run(
+      input: RDD[(Double, Double, Double)],
+      isotonic: Boolean = true): IsotonicRegressionModel = {
+    createModel(
+      parallelPoolAdjacentViolators(input, isotonic),
+      isotonic)
+  }
+
+  /**
+   * Creates isotonic regression model with given parameters
+   *
+   * @param predictions labels estimated using isotonic regression algorithm.
+   *                    Used for predictions on new data points.
+   * @param isotonic isotonic (increasing) or antitonic (decreasing) sequence
+   * @return isotonic regression model
+   */
+  protected def createModel(
+      predictions: Array[(Double, Double, Double)],
+      isotonic: Boolean): IsotonicRegressionModel = {
+
+    val labels = predictions.map(_._1)
+    val features = predictions.map(_._2)
+
+    new IsotonicRegressionModel(features, labels)
+  }
+
+  /**
+   * Performs a pool adjacent violators algorithm (PAVA)
+   * Uses approach with single processing of data where violators
+   * in previously processed data created by pooling are fixed immediatelly.
+   * Uses optimization of discovering monotonicity violating sequences (blocks)
+   * Method in situ mutates input array
+   *
+   * @param in input data
+   * @param isotonic asc or desc
+   * @return result
+   */
+  private def poolAdjacentViolators(
+      in: Array[(Double, Double, Double)],
+      isotonic: Boolean): Array[(Double, Double, Double)] = {
+
+    // Pools sub array within given bounds assigning weighted average value to all elements
+    def pool(in: Array[(Double, Double, Double)], start: Int, end: Int): Unit = {
+      val poolSubArray = in.slice(start, end + 1)
+
+      val weightedSum = poolSubArray.map(lp => lp._1 * lp._3).sum
+      val weight = poolSubArray.map(_._3).sum
+
+      var i = start
+      while (i <= end) {
+        in(i) = (weightedSum / weight, in(i)._2, in(i)._3)
```

Comment:
My previous comment about data compression still applies. For each pool, we only need to store the sum of labels and the sum of weights, and the prediction is the weighted average.